starlark/interp.go

```diff
@@ -85,6 +87,15 @@ func (fn *Function) CallInternal(thread *Thread, args Tuple, kwargs []Tuple) (Va
 	code := f.Code
 loop:
 	for {
+		thread.Steps++
+		if thread.Steps == 0 {
+			thread.Cancel("too many steps")
+		}
+		if reason := atomic.LoadPointer((*unsafe.Pointer)(unsafe.Pointer(&thread.cancelReason))); reason != nil {
```

Comment:
The atomic load, compare, and branch on every iteration seems expensive.

`Cancel` could set `Steps` to `^0`, but I suppose an atomic load would still be needed.

That could be optimized by keeping a local step count and updating the thread's step count on certain instructions (on builtin, return, or back branch) or when the local step count crosses some constant threshold.

No need to optimize that now, but it may require changing the contract of `Steps` in the future. Better to keep it abstract or vague.

Reply:
I measured it on x86 and it wasn't noticeable (sorry I didn't keep the data). The address is always in L1 cache, the branch is extremely predictable, and on Intel at least, every load is an atomic load. (The atomic load is much cheaper than selecting on the Done channel of a context: even though fundamentally that too boils down to an atomic load, compare, and branch, the function call overheads are significant.)

I considered using a local variable but it complicates the logic, and I suspect it could increase register pressure in the rest of the bytecode interpreter.